src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1072 @@
+
+                self.count = response.count + 1
+                self.first_timestamp = response.first_timestamp
+                try:
+                    self.api.patch_namespaced_event(self.event_name, self.namespace, self.event_body)
+                except ApiException as e:
+                    self.api_status = e.status
+                else:
+                    self.api_status = 200
+
+        else:
+            self.api_status = 200
+
+    @property
+    def api_success(self):
+        return self.api_status == 200
+
+    def update(self, log_entry):
+        self.message = log_entry.event_msg
+        self.event_type = log_entry.event_type
+        self.last_timestamp = datetime.now(timezone.utc)
+        self.count += 1
+
+        try:
+            self.api.patch_namespaced_event(self.event_name, self.namespace, self.event_body)
+        except ApiException as e:
+            if e.status == 404:
+                # tried to patch 404 indicates it's TTL has expired
+                try:
+                    self.api.create_namespaced_event(self.namespace, self.event_body)
+                except ApiException as e:
+                    self.api_status = e.status
+                else:
+                    self.api_status = 200
+        else:
+            self.api_status = 200 
+
+
+class EventProcessor(BaseThread):
+    """Handle a global queue used to track events we want to send/update to kubernetes"""
+
+    can_run = True
+
+    def __init__(self, config_watcher, event_retention_days):
+        super(EventProcessor, self).__init__()
+        self.events = dict()
+        self.config_watcher = config_watcher
+        self.event_retention_days = event_retention_days
+
+    def startup(self):
+        """Log an event to show we're active"""
+        
+        event = KubernetesEvent(
+            LogEntry(
+                source='self',
+                msg='Ceph log -> event tracking started',
+                msg_type='startup',
+                level='INF',
+                tstamp=None
+            ),
+            unique_name=False
+        )
+        
+        event.write()
+        return event.api_success
+    
+    @property
+    def ok(self):
+        return self.startup()
+
+    def prune_events(self):
+
+        oldest = datetime.now(timezone.utc) - timedelta(days=self.event_retention_days)
+        local_events = dict(self.events)
+
+        for event_name in sorted(local_events,
+                                 key = lambda name: local_events[name].last_timestamp):
+            event = local_events[event_name]
+            if event.last_timestamp >= oldest:
+                break
+            else:
+                # drop this event
+                log.debug("[DBG] removing old event : {}".format(event_name))
+                del self.events[event_name]
+
+    def process(self, log_object):
+        
+        log.debug("[DBG] k8sevents processing : {}".format(str(log_object)))
```

Comment:
Do you need the `[DBG]` prefix if you're calling `log.debug()`?

Reply:
debug sets the level, but the inclusion of [DBG] in the message makes entries easier to find when you turn on verbosity in the mgr

Reply:
What is your use case here? You can grep for `k8sevents` in the mgr log file.

Reply:
Use case? When I look at the mgr log this pattern is already being used
log_channel(cluster) log [DBG] : pgmap v273: 0 pgs: ; 0 B data, 5.4 MiB used, 3.0 TiB / 3.0 TiB avail
log_channel(audit) log [DBG] : from='client.24287 -' entity='client.admin' cmd=[{"prefix": "dashboard create-self-signed-cert", "target": ["mgr", ""], "format": "json"}]: dispatch

I think this is set in src/common/LogEntry.h

Since k8s is using this stream as a data source, I was using this example. Yes I can grep the logfile for k8sevents and see everything, but having the DBG or INF in the message provides additional granularity. Was LogEntry a bad example?

Reply:
We're using [`dout()`](https://docs.ceph.com/docs/master/rados/troubleshooting/log-and-debug/#subsystem-log-and-debug-settings) to write Python log messages to the mgr log file. Which uses log levels between 0 and 20. 

There is a mapping between The python levels and the ceph level:

https://github.com/ceph/ceph/blob/122c17f55fad81b9c868460b959126b5d99e2eff/src/pybind/mgr/mgr_module.py#L55-L63

If we want to additionally add the python log to the log messages, I think

https://github.com/ceph/ceph/blob/122c17f55fad81b9c868460b959126b5d99e2eff/src/pybind/mgr/mgr_module.py#L68

would be a better place for that.

The cluster and audit log channels work slightly different that the debug logs.

Reply:
Ok  - interesting discussion. @sebastian-philipp  Just to clarify, are you saying that for merge you want me to remove the [INF]/[DBG] content from the log statements in the module?

Reply:
I think it would create less confusion, if you would leave them out.

Reply:
ACK